src/packet/buffer_rx.rs

```diff
@@ -146,49 +148,132 @@ impl DepacketizingBuffer {
         //     self.segments
         // );
 
-        let (start, stop) = self.segments.first()?;
+        let (start, stop) = *self.segments.first()?;
+
+        let seq = {
+            let last = self.queue.get(stop).expect("entry for stop index");
+            last.meta.seq_no
+        };
+
+        // depack ahead to check contiguity, even if we may not emit right away
+        let dep = match self.depacketize(start, stop, seq) {
+            Ok(d) => d,
+            Err(e) => {
+                // this segment cannot be decoded correctly
+                // remove from the queue and return the error
+                self.last_emitted = Some((seq, CodecExtra::None));
+                self.queue.drain(0..=stop);
+                return Some(Err(e));
+            }
+        };
+
+        let contiguous = 'contig: {
+            if self.is_following_last(start) {
+                break 'contig true;
+            }
+
+            let Some((last_seq, last_codec_extra)) = self.last_emitted else {
+                break 'contig true;
+            };
+
+            const VP8_DEPACK_OPTIMIZATION: bool = true;
+
+            match (last_codec_extra, dep.codec_extra) {
+                (CodecExtra::Vp8(prev), CodecExtra::Vp8(next)) => {
+                    if !VP8_DEPACK_OPTIMIZATION {
```

Comment:
Maybe we extract this VP8 specific stuff to a function or make this whole thing polymorphic based on the codec, both the cache and this extra parsing is codec specific after all

Reply:
I did think about delegating to isolated codec-specific logic, maybe through a trait .. but to be honest I would do that when we support 2 or 3 codecs like this. At the moment we are not even sure this is a good idea long term

Reply:
Perhaps you are right, a function in the interim?

Reply:
okie